### PR TITLE
(ASC-1503) update test for validating volume snapshots

### PR DESCRIPTION
**Pull request by phongdly**
Prior to this PR, volumes are not created (due to the recent changes
in converge steps) and volume snapshot is not created for the test.

This PR adds volume and volume snapshot creation so that the test
can really test the cinder lvm volumes and volume snapshots.

### CODE REVIEW

File: molecule/default/tests/test_cinder_lvm.py [hunk 1]
```diff
@@ -27,14 +31,36 @@ def test_cinder_lvs_volume_on_node(os_api_conn, host):
...
     # get SnapshotDetails
-    snapshot_details = os_api_conn.block_storage.snapshots()
 
+    # create volume snapshot
     for vol in vol_details:
+        vol_id = str(vol.to_dict()['id'])
+        snapshot_name = "test_volume_snapshot{}".format(
+            str(helpers.generate_random_string()))
+
+        os_api_conn.create_volume_snapshot(volume_id=vol_id,
```
Comment (johnduarte) [inline molecule/default/tests/test_cinder_lvm.py]: Do you want to snapshot all of the volumes or just the one that was created earlier?  It appears that this loop attempts to create snapshots of all volumes.  This seems like a waste of resources and increases the risk of a spurious failure not related to the volume that you are specifically using in this test.

Should this be checking the status of the volume prior to attempting the snapshot?  Are there statuses that would prevent the snapshot from being created?  Should such a check be added as an assertion prior to the attempt to create the snapshot?

File: molecule/default/tests/test_cinder_lvm.py [hunk 2]
```diff
@@ -52,12 +78,21 @@ def test_cinder_lvs_volume_on_node(os_api_conn, host):
...
-                    assert host.run(cmd), "The command: '{}' failed to run on" \
-                                          "host '{}'.".format(cmd, current_host)
+                    assert host.run(cmd).rc == 0, \
+                        "The command: '{}' failed to run on" \
+                        "host '{}'.".format(cmd, current_host)
+
+    # tear down
+    # tear down snapshot only, the volume tear down is handled in create_volume
+    # factory function
+    for snapshot in os_api_conn.block_storage.snapshots():
```
Comment (johnduarte) [inline molecule/default/tests/test_cinder_lvm.py]: Wouldn't this destroy *all* snaphots not just the one(s) created as part of this test?  The teardown should limit itself to the resources created during the test.  Should there also be a teardown to remove the volume created for the test as well?

File: molecule/default/tests/test_cinder_lvm.py (hunk 1 above)
Comment (phongdly) [reply molecule/default/tests/test_cinder_lvm.py]: For all system tests that need to create a volume, we are using pytest fixture `create_volume` in pytest-rpc where volume is created with a `yield` statement instead of `return`, therefore the teardown code in that fixture will be executed upon the exiting of each test case, all volumes should be deleted before entering this test. The number of volumes created in this test depends on how many cinder hosts in , so two volumes and two volume snapshot will be created. I think it would be ok to snapshot two volumes instead of one. 

The volume status is actually checked in the pytest fixture `create_volume` as well so we don't have to check it again here.

File: molecule/default/tests/test_cinder_lvm.py (hunk 2 above)
Comment (phongdly) [reply molecule/default/tests/test_cinder_lvm.py]: Same as my above comment, there will be two snapshots and the volumes will be destroyed at the end of the test.

File: molecule/default/tests/test_cinder_lvm.py (hunk 1 above)
Comment (johnduarte) [reply molecule/default/tests/test_cinder_lvm.py]: @phongdly thanks for the clarification.  Assuming that all volumes are created by the `create_volume` fixture seems like a risk to me.  What if openstack-ansible-ops decides to create volumes as part of its deploy, or I erroneously commit a test that doesn't use the fixture?  These are obviously edge cases, but I see them as risks just the same.  I am fine merging this PR if you would do me the courtesy of creating a tech-debt ticket to capture the risk that this loop has the potential to encounter volumes that the test was not responsible for creating.

File: molecule/default/tests/test_cinder_lvm.py [hunk 3]
```diff
@@ -27,14 +31,36 @@ def test_cinder_lvs_volume_on_node(os_api_conn, host):
     Args:
         os_api_conn (openstack.connection.Connection): An authorized API
             connection to the 'default' cloud on the OpenStack infrastructure.
+        create_volume (def): A factory function for generating volumes.
+        openstack_properties (dict): OpenStack facts and variables from Ansible
         host (testinfra.host.Host): Testinfra host fixture.
     """
+    # create volume:
+    create_volume(
```
Comment (ryan-rs) [inline molecule/default/tests/test_cinder_lvm.py]: You can capture the volume details by just assigning a variable to the `create_volume` call:
```suggestion
    test_volume = create_volume(
```

**Comment by phongdly** [comment]
@johnduarte @ryan-rs : updated!

File: molecule/default/tests/test_cinder_lvm.py [hunk 4]
```diff
@@ -52,12 +70,20 @@ def test_cinder_lvs_volume_on_node(os_api_conn, host):
...
-                                          "host '{}'.".format(cmd, current_host)
+                    assert host.run(cmd).rc == 0, \
+                        "The command: '{}' failed to run on" \
+                        "host '{}'.".format(cmd, current_host)
+
+    # tear down
+    # tear down snapshot only, the volume tear down is handled in create_volume
+    # factory function
+    snapshot_id = test_snapshot.id
+    os_api_conn.delete_volume_snapshot(name_or_id=snapshot_id, wait=True)
```
Comment (ryan-rs) [inline molecule/default/tests/test_cinder_lvm.py]: Given that this is a teardown step I would suggest creating a fixture that guarantees teardown to happen regardless of test outcome.

Comment (phongdly) [reply molecule/default/tests/test_cinder_lvm.py]: @ryan-rs I thought about creating fixture for create_snapshot but then this feature teardown must be executed before the teardown of create_volume fixture, otherwise it will fail because the volume has associated volume snapshot. 
Any suggestion for easily and reliably setting the order of executing teardowns when there are more than one fixtures used in the test case?

Comment (ryan-rs) [reply molecule/default/tests/test_cinder_lvm.py]: @phongdly I created a PR (#224) that uses a snapshot fixture. Given how fixtures work the order of teardown should be correct with the snapshot getting deleted first followed by the volume getting deleted. 

I couldn't test the change because I don't have any operational environments in Phobos. Let me know if it works or not.

Comment (phongdly) [reply molecule/default/tests/test_cinder_lvm.py]: @ryan-rs ah I understand what you meant by the fixture which includes both creating volume and volume snapshot in one fixture. 

BTW, the above PR works!
